crates/amaru/src/stages/consensus/forward_chain/tcp_forward_chain_server.rs

```diff
@@ -0,0 +1,211 @@
+    #[expect(clippy::unwrap_used)]
+    pub fn create(
+        store: Arc<dyn ChainStore<Header>>,
+        runtime: Arc<AcTokio>,
+        tcp_listener: TcpListener,
+        network_magic: u64,
+        max_peers: usize,
+        our_tip: HeaderTip,
+    ) -> anyhow::Result<Self> {
+        let clients = runtime
+            // FIXME: This is a temporary stop gap solution while we wait
+            // to refactor to use pure-stage. Acto library has as a capped
+            // size for the mailbox and drops incoming messages when it's
+            // full.  This should not be a problem in real life, but while
+            // we are syncing _and_ forwarding at the same time for demo
+            // purpose, this is problematic.
+            .with_mailbox_size(1_000_000)
+            .spawn_actor("chain_forward", |cell| {
+                client_supervisor(cell, store.clone(), max_peers)
+            })
+            .me;
+
+        let our_tip = Arc::new(Mutex::new(our_tip.clone()));
+        runtime.with_rt(|rt| {
+            let our_tip_clone = our_tip.clone();
+            let clients_clone = clients.clone();
+            rt.spawn(async move {
+                loop {
+                    // due to the signature of TcpListener::accept, this is the only way to use this API
+                    // in particular, it isn’t possible to poll for new peers within the `schedule` method
+                    match PeerServer::accept(&tcp_listener, network_magic).await {
+                        Ok(peer) => {
+                            let our_tip = our_tip_clone.lock().unwrap().clone();
+                            clients_clone.send(ClientMsg::Peer(
+                                peer,
+                                Tip(our_tip.point().pallas_point(), our_tip.block_height()),
+                            ));
+                        }
+                        Err(e) => {
+                            tracing::warn!(
+                                target: EVENT_TARGET,
+                                "error accepting peer: {e}",
+                            );
+                            continue;
+                        }
+                    };
+                }
+            })
+        });
+
+        Ok(Self {
+            _runtime: runtime,
+            clients: clients.clone(),
+            our_tip: our_tip.clone(),
+        })
+    }
+}
+
+/// This implementation of ForwardEventListener sends the received events to all connected clients.
+#[async_trait]
+impl ForwardEventListener for TcpForwardChainServer {
+    async fn send(&self, event: ForwardEvent) -> anyhow::Result<()> {
+        match event {
+            ForwardEvent::Forward(header) => {
+                let mut our_tip = self
+                    .our_tip
+                    .lock()
+                    .map_err(|e| anyhow::anyhow!("Mutex poisoned: {}", e))?;
+                *our_tip = header.as_header_tip();
```

Comment:
probably irrelevant in this case but keeping this mutex locked after this line feels dirty

Reply:
Fixed.